v2/stream/new/flow.go

```diff
@@ -0,0 +1,193 @@
+	defer cancelWriteSub()
+	if err := c.writeSub(ctxWithTimeoutWriteSub); err != nil {
+		return err
+	}
+
+	ctxWithTimeoutReadSub, cancelReadSub := context.WithTimeout(ctx, initializeTimeout)
+	defer cancelReadSub()
+	if err := c.readSubResponse(ctxWithTimeoutReadSub); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var ErrNoConnected = errors.New("did not receive connected message")
+
+func (c *client) readConnected(ctx context.Context) error {
+	b, err := c.conn.readMessage(ctx)
+	if err != nil {
+		return err
+	}
+	var resps []struct {
+		T   string `msgpack:"T"`
+		Msg string `msgpack:"msg"`
+	}
+	if err := msgpack.Unmarshal(b, &resps); err != nil {
+		return err
+	}
+	if len(resps) != 1 {
+		return ErrNoConnected
+	}
+	if resps[0].T != "success" || resps[0].Msg != "connected" {
+		return ErrNoConnected
+	}
+	return nil
+}
+
+func (c *client) writeAuth(ctx context.Context) error {
+	msg, err := msgpack.Marshal(map[string]string{
+		"action": "auth",
+		"key":    c.key,
+		"secret": c.secret,
+	})
+	if err != nil {
+		return err
+	}
+
+	return c.conn.writeMessage(ctx, msg)
+}
+
+var ErrBadAuthResponse = errors.New("did not receive authenticated message")
+
+func (c *client) readAuthResponse(ctx context.Context) (canBeRetried bool, err error) {
+	b, err := c.conn.readMessage(ctx)
+	if err != nil {
+		return false, err
+	}
+	var resps []struct {
+		T    string `msgpack:"T"`
+		Msg  string `msgpack:"msg"`
+		Code int    `msgpack:"code"`
+	}
+	if err := msgpack.Unmarshal(b, &resps); err != nil {
+		return false, err
+	}
+	if len(resps) != 1 {
+		return false, ErrBadAuthResponse
+	}
+
+	// A previous connection may be "stuck" on the server so we may run into
+	// `[{"T":"error","code":406,"msg":"connection limit exceeded"}]`
+	if resps[0].T == "error" {
+		err := fmt.Errorf("auth: error from server: %s", resps[0].Msg)
+		if resps[0].Code == 406 {
+			return true, err
+		}
+		return false, err
+	}
+	if resps[0].T != "success" || resps[0].Msg != "authenticated" {
+		return false, ErrBadAuthResponse
+	}
+
+	return false, nil
+}
+
+func (c *client) writeSub(ctx context.Context) error {
+	msg, err := getSubChangeMessage(true, c.trades, c.quotes, c.bars)
+	if err != nil {
+		return err
+	}
+
+	return c.conn.writeMessage(ctx, msg)
+}
+
+var ErrSubResponse = errors.New("did not receive subscribed message")
+
+func (c *client) readSubResponse(ctx context.Context) error {
+	b, err := c.conn.readMessage(ctx)
+	if err != nil {
+		return err
+	}
+	var resps []struct {
+		T      string   `msgpack:"T"`
+		Msg    string   `msgpack:"msg"`
+		Code   int      `msgpack:"code"`
+		Trades []string `msgpack:"trades"`
+		Quotes []string `msgpack:"quotes"`
+		Bars   []string `msgpack:"bars"`
+	}
+	if err := msgpack.Unmarshal(b, &resps); err != nil {
+		return err
+	}
+	if len(resps) != 1 {
+		return ErrSubResponse
+	}
+
+	if resps[0].T == "error" {
+		return fmt.Errorf("sub: error from server: %s", resps[0].Msg)
+	}
+	if resps[0].T != "subscription" {
```

Comment:
What happens if we get data (for example a trade from a previous subscription) instead of the subscription confirmation here?

Reply:
Although the naming may be a bit confusing, functions in `flow.go` are only called by `initialize`: this is when we connect to the server, authenticate and set up initial subscriptions.
In this flow no other messages are acceptable.

Reply:
It's clear now. Sorry. 🔫🔫🔫